### ISSUE TITLE
`mz_hydration_statuses` is wrong for REFRESH MVs

### ISSUE THREAD

**Issue by ggevay** [MaterializeInc/materialize#25518]
### What version of Materialize are you using?

3976a7026d4c4c7a131f26bcec9c110141d362bc

### What is the issue?

@teskje: It occurs to me that the hydrated flag is not very useful for REFRESH MVs. It will flip to true when the persist_sink operator reports frontier advancement beyond the dataflow as_of. Normally that means all upstream operators have processed the as_of time (i.e. the snapshot), but I believe for REFRESH MVs we bump the write frontier to the next refresh interval immediately.

Repro:
```
CREATE CLUSTER c1 SIZE '1', REPLICATION FACTOR 1;
SET CLUSTER = 'c1';

CREATE TABLE t (a int);

INSERT INTO t VALUES (10000000);

CREATE MATERIALIZED VIEW mv WITH (REFRESH EVERY '1 day') AS
SELECT count(*) FROM (SELECT generate_series(1,a) FROM t);
```
then wait for the first refresh, and then restart envd or the replica. The CPU will be spinning for several seconds even after
```
SELECT hydrated
FROM mz_internal.mz_compute_hydration_statuses h JOIN mz_objects o ON (h.object_id = o.id)
WHERE name = 'mv';
```
shows that it's hydrated.

Btw. currently the hydrated flag is fine when creating a new MV, because we set the `as_of` to the first refresh. But this will also break with the planned fix for #25278...

(Note that, in contrast to normal MVs, a REFRESH MV is usually readable already even before it's rehydrated, unless the hydration straddles a refresh or it's the first hydration.)

**Comment by teskje** [comment]
> (Note that, in contrast to normal MVs, a REFRESH MV is usually readable already even before it's rehydrated, unless the hydration straddles a refresh.)

Wait, that's also true for normal MVs, no? It's also true for sources, and generally all persist collections.

**Comment by ggevay** [comment]
I have an idea: We could tweak the SQL-level definition of `mz_hydration_statuses` to have a special logic for MVs with any REFRESH options: instead of looking at `mz_compute_hydration_statuses`, it would look at `mz_compute_operator_hydration_statuses` and check that all operators have hydrated. @teskje, what do you think?

Note that currently we don't have a reliable way to know from internal tables whether an MV has REFRESH options, but I'm planning to add a column to `mz_materialized_views` anyway that would indicate this (https://github.com/MaterializeInc/materialize/issues/24051).

**Comment by teskje** [comment]
That seems like an okay quick fix, but then we'd still have the bug in `mz_compute_hydration_statuses`. And for sequential hydration we also need some way for the replica to figure out whether a dataflow is fully hydrated or not.

So I think we want to determine dataflow hydration on the replica and then send a `Status` response to the controller, like we do for operator hydration. Then we need to fix the issue only once in the replica code. And for that using operator hydration statuses sounds like a good idea.

**Comment by ggevay** [comment]
In the meantime, I postponed this to after the automated cluster scheduling.

**Comment by ggevay** [comment]
Noting here that [a user](https://github.com/MaterializeInc/accounts/issues/3) also encountered a case (twice) where a REFRESH MV was shown as not hydrated, when in fact it probably was. We don't know what's caused this. [Slack discussion.](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1708708936901689?thread_ts=1708622555.214129&cid=C02PPB50ZHS)

**Comment by ggevay** [comment]
Jan's different quick fix PR: https://github.com/MaterializeInc/materialize/pull/26807
